---
title: "babette demo"
author: "RichÃ¨l J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{babette demo}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

This vignette demonstrates how to use `babette`.

First, load the library:

```{r load_babette, results='hide', warning=FALSE, error=FALSE, message=FALSE}
library(babette)
```

The following example shows how to analyze one alignment:

 * How to let `babette` run BEAST2
 * How the plot the posterior estimates
 * How to show the effective sample sizes (ESS)
 
In all cases, this is done for a short MCMC chain length of 10K:

```{r}
sample_interval <- 1000
mcmc <- create_test_mcmc(
  chain_length = 10000,
  tracelog = create_tracelog(
    log_every = sample_interval
  )
)
inference_model <- create_test_inference_model(
  mcmc = mcmc
)
```

For a reliable inference, use an ESS of at least 200.

```{r}
if (is_beast2_installed()) {
  out <- bbt_run_from_model(
    fasta_filename = get_babette_path("anthus_aco_sub.fas"),
    inference_model = inference_model
  )
}
```

```{r}
library(ggplot2)
p <- ggplot(
  data = out$estimates,
  aes(x = Sample)
) 
p + geom_line(aes(y = TreeHeight), color = "green")
p + geom_line(aes(y = YuleModel), color = "red")
p + geom_line(aes(y = birthRate), color = "blue")

```

Effective sample sizes, with 20% burn-in removed:

```{r}
traces <- remove_burn_ins(
  traces = out$estimates, 
  burn_in_fraction = 0.2
)
esses <- t(calc_esses(traces, sample_interval = sample_interval))
colnames(esses) <- "ESS"
knitr::kable(esses)
```

Summary statistics:
  
```{r}
sum_stats <- t(calc_summary_stats(traces$posterior, sample_interval = sample_interval))
colnames(sum_stats) <- "Statistic"
knitr::kable(sum_stats)
```

Phylogenies:

```{r fig.width=7, fig.height=7}
plot_densitree(out$anthus_aco_sub_trees, width = 2)
```
